File: apps/commerce-api/src/main/java/com/loopers/domain/product/ProductService.java
package com.loopers.domain.product;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

import org.springframework.data.domain.Page;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

import com.loopers.support.error.CoreException;
import com.loopers.support.error.ErrorType;

import lombok.RequiredArgsConstructor;

@Component
@RequiredArgsConstructor
public class ProductService {
    private final ProductRepository productRepository;
    private final ProductCacheRepository productCacheRepository;
    private final ProductMapper productMapper;

    @Transactional(readOnly = true)
    public Product get(Long id) {
        return productRepository.find(id)
            .orElseThrow(() -> new CoreException(ErrorType.NOT_FOUND, "상품을 찾을 수 없습니다. 상품 ID: " + id));
    }

    @Transactional(readOnly = true)
    public ProductInfo getInfo(Long id) {
        Optional<ProductInfo> cachedInfo = productCacheRepository.getProductInfo(id);
        if (cachedInfo.isPresent()) {
            return cachedInfo.get();
        }
        Product product = productRepository.find(id)
            .orElseThrow(() -> new CoreException(ErrorType.NOT_FOUND, "상품을 찾을 수 없습니다. 상품 ID: " + id));
        ProductInfo info = ProductInfo.from(product);
        productCacheRepository.setProductInfo(id, info);
        return info;
    }

    @Transactional(readOnly = true)
    public List<Product> getAll(List<Long> ids) {
        return productRepository.findAll(ids);
    }

    @Transactional(readOnly = true)
    public Page<Product> getAll(ProductCommand.Search command) {
        return productRepository.findAll(command);
    }

    @Transactional(readOnly = true)
    public List<ProductPrice> getAvailableProductPrices(ProductCommand.GetAvailable command) {
        List<Long> productIds = command.options().stream()
            .map(option -> option.productId())
            .distinct()
            .toList();

        Map<Long, Product> products = productRepository.findAll(productIds)
            .stream()
            .collect(Collectors.toMap(
                Product::getId,
                product -> product));

        if (products.size() < productIds.size()) {
            throw new CoreException(ErrorType.NOT_FOUND, "구매할 상품이 존재하지 않습니다.");
        }

        return productMapper.mapToProductPrices(command.options(), products);
    }
}

File: apps/commerce-streamer/src/main/kotlin/com/loopers/domain/product/ProductService.kt
package com.loopers.domain.product

import com.loopers.support.error.CoreException
import com.loopers.support.error.ErrorType
import org.springframework.stereotype.Component

@Component
class ProductService(
    private val productRepository: ProductRepository,
    private val productEventPublisher: ProductEventPublisher,
    private val productCacheRepository: ProductCacheRepository
) {
    fun deduct(command: ProductCommand.Deduct) {
        val products = productRepository.findAll(command.productIds())
        command.options
            .forEach { option ->
                val product = products.find { it.id == option.productId } ?: throw CoreException(ErrorType.NOT_FOUND)
                product.deduct(option.productOptionId, option.quantity)
            }
        val soldOutProducts = products.filter { it.isSoldOut() }
        soldOutProducts.forEach { it.markAsSoldOut() }

        if (soldOutProducts.isNotEmpty()) {
            val soldOutProductIds = soldOutProducts.mapNotNull { it.id }
            productEventPublisher.publishSoldOutEvent(soldOutProductIds)
        }
    }

    fun deleteCaches(event: ProductSoldOutEvent) {
        event.productIds.forEach {
            productCacheRepository.deleteProductInfo(it)
        }
    }
}
